rust/processor/src/worker.rs

```diff
@@ -1182,3 +1188,184 @@ pub async fn create_fetcher_loop(
         }
     }
 }
+
+/// Gap detector task
+/// 1. Pull from gap detector channel
+/// 2. Checks for any gaps. We detect a gap when GAP_DETECTION_BATCH_COUNT batches are processed and there is no advancement of prev_end version
+pub async fn create_gap_detector_task(
+    mut receiver: tokio::sync::mpsc::Receiver<ProcessedVersions>,
+    db_pool: PgDbPool,
+    processor_name: String,
+    starting_version: u64,
+) {
+    // Keep track of the start versions we've seen
+    let mut seen_versions = HashMap::new();
+    // Keep track of the latest batch processed without gaps
+    let mut maybe_prev_batch: Option<ProcessedVersions> = None;
+    // Counter of how many batches have been processed with no gaps
+    let mut num_batches_processed_without_gap = 0;
+    // Counter of how many batches have been processed with a gap from prev_end
+    let mut num_batches_processed_with_gap = 0;
+
+    loop {
+        match receiver.recv().await {
+            Some(result) => {
```

Comment:
I think you could flatten this logic, instead of being so indented- i.e
```
  let result = match receiver.recv().await {
            Some(result) => result,
            None => continue,
}
```